### PR TITLE
[INLONG-5722][Agent] Support Redis Source

### LINKED ISSUE #5722: [Feature][Agent] Support Redis Source
### Description

Redis is an in-memory data structure store, used as a distributed, in-memory key–value database, cache and message broker, with optional durability.

### Use case

_No response_

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by iamsee123**
### Prepare a Pull Request

- Fixes #5722

### Motivation
make agent module support redis source

### Modifications

+ add RedisSource and RedisReader
+ add unit test for redisReader

### Verifying this change

*(Please pick either of the following options)*

- [ ] This change is a trivial rework/code cleanup without any test coverage.

- [ ] This change is already covered by existing tests, such as:
  *(please describe tests)*

- [x] This change added tests and can be verified as follows:
  - TestRedisReader

### CODE REVIEW

**Comment by dockerzhang** [comment]
@pocozh @EMsnap PTAL

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+    private String instanceId;
+    private String replId;
+    private String snapShot;
+    private boolean destroyed;
+    private Replicator redisReplicator;
+    private LinkedBlockingQueue<String> redisMessageQueue;
+    private boolean finished = false;
+    private ExecutorService executor;
+    private Gson gson;
+
```
Comment (GanfengTan) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:84]: Suggest to add filter

**Review by Oneal65 — APPROVED** [review]

**Comment by dockerzhang** [comment]
@iamsee123 please fix the comment and conficts

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java (hunk 1 above)
Comment (iamsee123) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:84]: what kind of filter

Comment (GanfengTan) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:84]: https://redis.io/docs/manual/programmability/eval-intro/
@iamsee123 I would like you support scripting with Lua, thanks answer to the question.

Comment (iamsee123) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:84]: By using lua-replicate-commands configuration directive and [redis.replicate_commands()](https://redis.io/topics/lua-api#redis.replicate_commands) Lua API can enable effects replication, then redisReader can read the command directly. However if using verbatim replication, we just read the eval command but the change of data can be read by rdb scan at the beginning.

Comment (GanfengTan) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:84]: OK

**Review by GanfengTan — APPROVED** [review]

**Review by pocozh — APPROVED** [review]

**Review by EMsnap — APPROVED** [review]

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            readerMetric.pluginReadFailCount.addAndGet(1);
+            LOGGER.error("Connect to redis {}:{} failed.", hostName, port);
+        }
+    }
+
+    private String getRedisUri() {
+        StringBuffer sb = new StringBuffer("redis://");
+        sb.append(hostName).append(":").append(port);
+        sb.append("?");
+        if (authPassword != null && !authPassword.equals("")) {
```
Comment (EMsnap) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:143]: StringUtils.isEmpty

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+                    }
+                })
+                .create();
+    }
+
+    /**
+     * init replicator's commandParser
+     */
+    private void initReplicator() {
+        redisReplicator.addCommandParser(CommandName.name("APPEND"), new DefaultCommandParser());
```
Comment (EMsnap) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:352]: can we reuse  DefaultCommandParser()  here ?

Comment (iamsee123) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java:352]: Each command has its own commandParser, so I have to change it into unified parser

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java (hunk 2 above)
Comment (iamsee123) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java]: fixed

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java (hunk 3 above)
Comment (iamsee123) [reply inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/RedisReader.java]: I my test we can reuse DefaultCommandParser().

**Pull request merged**